### PR TITLE
PYTHON-2634 Only update pools for data-bearing servers

### PR DESCRIPTION
**Pull request by ShaneHarvey**
Fixes a noisy OperationFailure: Authentication failed error.
Do not attempt to create unneeded connections to arbiters, ghosts,
hidden members, or unknown members.

### CODE REVIEW

File: test/test_client.py [hunk 1]
```diff
@@ -1944,5 +1945,61 @@ def timeout_task():
...
+        self.assertEqual(
+            listener.event_count(monitoring.ConnectionCreatedEvent), 2)
+        # Assert that we do not create connections to arbiters.
+        arbiter = c._topology.get_server_by_address(('c', 3))
+        self.assertFalse(arbiter.pool.sockets)
+        # Assert that we do not create connections to unknown servers.
+        arbiter = c._topology.get_server_by_address(('d', 4))
+        self.assertFalse(arbiter.pool.sockets)
+
+    def test_direct_client_maintains_pool_to_arbiter(self):
```
Comment (prashantmital) [inline test/test_client.py:1979]: In the `directConnectio=True` scenario, should we have any special handling for the case where authentication is enabled on the arbiter? Based on [these docs](https://docs.mongodb.com/manual/core/replica-set-arbiter/#authentication): 

> the only way to log on to an arbiter with authorization active is to use the localhost exception.

it seems that we will never be able to populate a pool that connects to an arbiter with auth, or am I reading that wrong? What will the user-facing behavior be in this case?

Comment (prashantmital) [inline test/test_client.py:1979]: In the `directConnectio=True` scenario, should we have any special handling for the case where authentication is enabled on the arbiter? Based on [these docs](https://docs.mongodb.com/manual/core/replica-set-arbiter/#authentication): 

> the only way to log on to an arbiter with authorization active is to use the localhost exception.

it seems that we will never be able to populate a pool that connects to an arbiter with auth, or am I reading that wrong? What will the user-facing behavior be in this case?

**Review by prashantmital — COMMENTED** [review]
One question about arbiters with auth - LGTM otherwise.

File: test/test_client.py (hunk 1 above)
Comment (ShaneHarvey) [reply test/test_client.py:1979]: Yes, if a user wants to connect directly to an arbiter they will need to remove auth parameters.

**Review by prashantmital — APPROVED** [review]

**Comment by ShaneHarvey** [comment]
Sigh, this change seems to make these two tests more flakey:
```
 [2021/04/07 19:05:07.562] FAIL [10.033s]: test_max_idle_time_reaper (test_client.TestClient)
 [2021/04/07 19:05:07.562] ----------------------------------------------------------------------
 [2021/04/07 19:05:07.562] Traceback (most recent call last):
 [2021/04/07 19:05:07.562]   File "/data/mci/0f97b1956ba2d463825378bfe9902bfe/src/test/test_client.py", line 473, in test_max_idle_time_reaper
 [2021/04/07 19:05:07.562]     "remove stale socket")
 [2021/04/07 19:05:07.562]   File "/data/mci/0f97b1956ba2d463825378bfe9902bfe/src/test/utils.py", line 741, in wait_until
 [2021/04/07 19:05:07.562]     raise AssertionError("Didn't ever %s" % success_description)
 [2021/04/07 19:05:07.562] AssertionError: Didn't ever remove stale socket
```
https://evergreen.mongodb.com/task/mongo_python_driver_tests_python_version_amazon1_test_ssl__platform~awslinux_auth~noauth_ssl~nossl_python_version~3.6_coverage~coverage_test_2.6_replica_set_patch_cc029a1e6208863eaab453777363d3935b927f32_606dfefb61837d0383bb0567_21_04_07_18_50_36

And:
```
 [2021/04/07 19:02:56.470] FAIL [10.028s]: test_min_pool_size (test_client.TestClient)
 [2021/04/07 19:02:56.470] ----------------------------------------------------------------------
 [2021/04/07 19:02:56.470] Traceback (most recent call last):
 [2021/04/07 19:02:56.470]   File "/data/mci/4b8f11bb525edfac5c311f757eda876c/src/test/test_client.py", line 519, in test_min_pool_size
 [2021/04/07 19:02:56.470]     "pool initialized with 10 sockets")
 [2021/04/07 19:02:56.470]   File "/data/mci/4b8f11bb525edfac5c311f757eda876c/src/test/utils.py", line 741, in wait_until
 [2021/04/07 19:02:56.470]     raise AssertionError("Didn't ever %s" % success_description)
 [2021/04/07 19:02:56.470] AssertionError: Didn't ever pool initialized with 10 sockets
```
https://evergreen.mongodb.com/task/mongo_python_driver_tests_python_version_amazon1_test_ssl__platform~awslinux_auth~noauth_ssl~nossl_python_version~3.6_coverage~coverage_test_4.2_replica_set_patch_cc029a1e6208863eaab453777363d3935b927f32_606dfefb61837d0383bb0567_21_04_07_18_50_36

I believe this is because the background thread now has to wait for the SDAM state to be discovered before attempting to populate the pool but I am very surprised to see that this wouldn't happen within the test's 10 second timeout. Perhaps there is too much thread contention and one of the monitors is being starved for 10 seconds?

**Review by prashantmital — APPROVED** [review]

**Pull request merged**